### ISSUE TITLE
Architecture Compliance Rule

### ISSUE THREAD

**Issue by deeprim** [arturbosch/detekt#1164]
Introduce new rule that will allow the one to prohibit usage of classes from a certain package in classes of an arbitrary package:

```yml
architecture-compliance:
  prohibit:
    in: 
      - 'com.mypackage.*'
      - 'com.myotherpackage.*'
    from:
      - 'org.unwantedpackage.*'
    message:
      - 'Classes from `org.unwantedpackage.*` are disallowed for usage within `com.mypackage`...'
    # the next is questionable, why we would reduce severity for prohibited class?
    severity: 'error'
```

The similar rule may be found in sonar rules: [Architecture Rule Engine](https://docs.sonarqube.org/display/SONARQUBE44/Architecture+Rule+Engine)

The rule will allow developers to ensure that for an example "web" tier won't sprawl into "service" tier (or "repository" tier)

**Comment by Mishkun** [comment]
This is not idiomatic for Kotlin. In Kotlin you can have the desired effect enforced by the compiler, using `internal` modifier with splitting your code in separate modules instead of packages (which are easy to break, anyway)

**Comment by deeprim** [comment]
@Mishkun, doesn't see how "idiomatic kotlin" may be applied here.
Say You develop api server using spring framework, how `internal` may help you to prohibit to use spring mvc classes in your repository tier?

**Comment by deeprim** [comment]
Updated description to denote that "unwanted" package is 3rd party code.

**Comment by egor-n** [comment]
If `com.mypackage` and `com.myotherpackage` are in their own modules then the ForbiddenImport could be used for this.

**Comment by schalkms** [comment]
Do we really want this in detekt? For me this seems like a use case for a custom rule.
Sonar also deprecated this rule for a reason.

[Reference](https://jira.sonarsource.com/browse/SONARJAVA-1774)
